File: CSharp/CubeSolver/Tools/BenchmarkRunner.cs
using System.Text.Json;
using CubeSolver.Utils;
using NetCubeSolver;
using NetCubeSolver.Benchmarking;
using NetCubeSolver.ExactCover;

namespace CubeSolver.Tools;

public class BenchmarkRunner: ITool
{
    
    private BenchmarkSettings _benchmarkSettings;
    private BenchmarkResults _benchmarkResults;
    private List<Config> _configs;
    
    
    public string Name => "Run Benchmark";

    public bool Run()
    {
        _benchmarkSettings = ProgramUtils.GetBenchmarkSettingsFromUser();
        _benchmarkResults = new BenchmarkResults();
        LoadConfigs();

        ISolver solver;
        if (_benchmarkSettings.MultiThreaded)
        {
            solver = new ThreadedCubeSolver();
        }
        else
        {
            solver = new CubeSolver();
        }

        foreach (var benchmark in _benchmarkSettings.Configurations)
        {
            var config = GetConfig(benchmark);
            if (config == null)
            {
                ProgramUtils.WriteFilledLine('!', $"Error: No config with id {benchmark.ConfigId} found");
                return false;
            }
            
            var threadCount = GetThreadCount(benchmark);
            
            ProgramUtils.WriteFilledLine('=', $"Running benchmark for {config.DisplayName}, Threaded: {_benchmarkSettings.MultiThreaded}, Threads: {threadCount}, Depth: {benchmark.InitialDepth}");
            var benchmarkResult = new BenchmarkResult()
            {
                ConfigId = config.Id,
                ConfigName = config.DisplayName,
                InitialDepth = _benchmarkSettings.MultiThreaded ? benchmark.InitialDepth : 0,
                Threads = threadCount
            };

            ProgramUtils.Indent();
            for (int i = 0; i < _benchmarkSettings.Iterations; i++)
            {
                solver.Solve(config, benchmark.InitialDepth, threadCount, false, false);
                benchmarkResult.RunDurationsMs.Add(solver.SolveTime);
                ProgramUtils.WriteFilledLine('.', $"Iteration {i + 1} of {_benchmarkSettings.Iterations} Took: {solver.SolveTime:n0}ms");
            }
            ProgramUtils.Unindent();
            benchmarkResult.TotalDurationMs = benchmarkResult.RunDurationsMs.Sum();
            benchmarkResult.AverageRunDurationMs = benchmarkResult.TotalDurationMs / benchmarkResult.RunDurationsMs.Count;
            benchmarkResult.ShortestRunDurationMs = benchmarkResult.RunDurationsMs.Min();
            benchmarkResult.LongestRunDurationMs = benchmarkResult.RunDurationsMs.Max();
            _benchmarkResults.Results.Add(benchmarkResult);
            ProgramUtils.WriteFilledLine('=', benchmarkResult.GetSummary());
            ProgramUtils.Write();
        }
        
        var resultsFile = Path.Combine(ProgramUtils.OutputDirectory, $"BenchmarkResultsCSharp-{_benchmarkSettings.Id}.json");
        var writeOptions = new JsonSerializerOptions { WriteIndented = true };
        File.WriteAllText(resultsFile, JsonSerializer.Serialize(_benchmarkResults, writeOptions));
        
        ProgramUtils.Write();
        ProgramUtils.WriteFilledLine('=', "Benchmarking Complete:");
        ProgramUtils.Indent();
        foreach (var benchmarkResult in _benchmarkResults.Results)
        {
            ProgramUtils.Write(benchmarkResult.GetSummary());
        }
        ProgramUtils.Unindent();
        
        
        
        return true;
    }

    private void LoadConfigs()
    {
        _configs = new List<Config>();
        DirectoryInfo dir = new DirectoryInfo(ProgramUtils.WorkingDirectory);
        var configFiles = dir.GetFiles().Where(f => f.Extension == ".json").ToList();
        var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
        foreach (var configFile in configFiles)
        {
            var config = JsonSerializer.Deserialize<Config>(File.ReadAllText(configFile.FullName), options);
            config.InitializePieces();
            config.Verbose = false;
            _configs.Add(config);
        }
    }


    private Config GetConfig(BenchmarkConfig benchmarkConfig)
    {
        foreach (var config in _configs)
        {
            if (config.Id == benchmarkConfig.ConfigId) return config;
        }
        
        return null;
    }

    private int GetThreadCount(BenchmarkConfig benchmarkConfig)
    {
        if (!_benchmarkSettings.MultiThreaded)
        {
            return 1;
        }
        return benchmarkConfig.MaxThreadsOverride > 0? benchmarkConfig.MaxThreadsOverride : _benchmarkSettings.MaxThreads;
    }
}

File: Kotlin/app/src/main/kotlin/ca/jwolf/cubesolver/tools/BenchmarkRunner.kt
package ca.jwolf.cubesolver.tools

import ca.jwolf.cubesolver.benchmarking.BenchmarkConfig
import ca.jwolf.cubesolver.benchmarking.BenchmarkResult
import ca.jwolf.cubesolver.benchmarking.BenchmarkResults
import ca.jwolf.cubesolver.benchmarking.BenchmarkSettings
import ca.jwolf.cubesolver.configs.PuzzleConfig
import ca.jwolf.cubesolver.datamodels.Solver
import ca.jwolf.cubesolver.utils.ProgramUtils
import ca.jwolf.cubesolver.utils.ProgramUtils.Companion.ConfigurationDirectory
import ca.jwolf.cubesolver.utils.ProgramUtils.Companion.formatWithCommas
import ca.jwolf.cubesolver.utils.ProgramUtils.Companion.jsonMapper
import com.fasterxml.jackson.module.kotlin.readValue
import kotlinx.coroutines.runBlocking
import java.io.File
import java.nio.file.Paths

class BenchmarkRunner: Tool {

    private lateinit var benchmarkSettings: BenchmarkSettings
    private lateinit var benchmarkResults: BenchmarkResults
    private lateinit var puzzleConfigs : List<PuzzleConfig>


    override fun toolName(): String {
       return "Run Benchmark"
    }

    override fun run(): Boolean {

        benchmarkSettings = ProgramUtils.getBenchmarkSettingsFromUser()
        benchmarkResults = BenchmarkResults()
        loadPuzzleConfigs()

        val solver: Solver
        if (benchmarkSettings.multiThreaded)
        {
            solver = ThreadedSolver()
        }
        else
        {
            solver = SingleThreadedSolver()
        }

        benchmarkSettings.configurations.forEach { config ->
            val puzzleConfig = puzzleConfigs.find { pConfig -> pConfig.id == config.configId }

            if (puzzleConfig == null)
            {
                ProgramUtils.writeFilledLine('!', "Error: No puzzle config found for id: ${config.configId}")
                return false
            }

            val threadCount = getMaxThreads(config)
            ProgramUtils.writeFilledLine('=', "Running benchmark for config: ${puzzleConfig.displayName}, Threaded: ${benchmarkSettings.multiThreaded}, Threads: $threadCount, Depth: ${config.initialDepth}")

            val benchmarkResult = BenchmarkResult(
                puzzleConfig.id,
                puzzleConfig.displayName,
                threadCount,
                if (benchmarkSettings.multiThreaded) config.initialDepth else 0
            )

            ProgramUtils.indent()
            for(i in 0 until benchmarkSettings.iterations)
            {
                // we want to block here, only 1 iteration should run at a time
                runBlocking {
                    solver.solve(puzzleConfig, config.initialDepth, threadCount, false, false)
                }
                benchmarkResult.runDurationsMs.add(solver.solveTimeMs())
                ProgramUtils.writeFilledLine('.', "Iteration ${i + 1} of ${benchmarkSettings.iterations} took: ${solver.solveTimeMs().formatWithCommas()} ms")

            }
            ProgramUtils.unindent()
            benchmarkResult.totalDurationMs = benchmarkResult.runDurationsMs.sum()
            benchmarkResult.averageRunDurationMs = benchmarkResult.totalDurationMs / benchmarkResult.runDurationsMs.size
            benchmarkResult.shortestRunDurationMs = benchmarkResult.runDurationsMs.min()
            benchmarkResult.longestRunDurationMs = benchmarkResult.runDurationsMs.max()
            benchmarkResults.results.add(benchmarkResult)
            ProgramUtils.writeFilledLine('=', benchmarkResult.getSummary())
            ProgramUtils.writeLine()

        }

        val results = jsonMapper.writeValueAsBytes(benchmarkResults)
        File(Paths.get(ProgramUtils.OutputDirectory, "BenchmarkResultsKotlin-${benchmarkSettings.id}.json").toString()).writeBytes(results)

        return true
    }

    private fun loadPuzzleConfigs()
    {
        val fileList = Paths.get(ConfigurationDirectory).toFile().listFiles()!!.map { it.name }.filter { it.endsWith(".json") }
        puzzleConfigs = fileList.map { jsonMapper.readValue<PuzzleConfig>(Paths.get(ConfigurationDirectory, it).toFile().readBytes().decodeToString()) }
        puzzleConfigs.forEach{
            it.initializePieces()
            it.verbose = false
        }
    }

    private fun getMaxThreads(benchmarkConfig: BenchmarkConfig) : Int
    {
        if (!benchmarkSettings.multiThreaded)
        {
            return 1
        }

        return if (benchmarkConfig.maxThreadsOverride > 0) benchmarkConfig.maxThreadsOverride else benchmarkSettings.maxThreads
    }
}